Fix window operation with kRange frames with column boundary

kagamiori opened the pull request:
Summary:
When Window operator is initialized, WindowBuild reorders the input columns to 
place partition keys and sorting keys first before adding input rows to 
RowContainer. However, for window operations that use kRange frame with 
column boundary, e.g., range between c1 preceding and current row, the original 
column index of the frame column before reordering is passed to 
computeKRangeFrameBounds(), which leads to incorrect result. This diff fixes 
this bug by passing the actual index of the frame column after the reordering to 
computeKRangeFrameBounds().

This diff fixes https://github.com/facebookincubator/velox/issues/9229.

Differential Revision: D55351848

kagamiori commented:
Hi @aditi-pandit, could you help review this fix? Thanks!

aditi-pandit commented:
@kagamiori : So your original idea to pass windowBuild_.inputType_  to createWindowFrame is actually a more correct fix imo. 

The reason you had other issues is because the wrong type is used to figure indices in multiple places...in sending the function parameter indices to the WindowFunction, in mapping output columns to input columns for getOutput().. There might be other places as well.

We likely didn't encounter these issues because none of our testing flipped order of columns in this way :( So we should incorporate this in the fuzzer as well.

Let me try your test with some fix logic. I'll send you a sample PR in mind in a few hours.

aditi-pandit commented:
@kagamiori : I debugged this a bit more. 

So there wasn't an issue with the function column indices or output indices because the WindowBuild code had constructed WindowPartitions (used for function and output indices) providing a column mapping using the input type order. 

The range frame code needed to use that mapping as well. Since the Window operator code uses the input type indices everywhere its consistent to send the original index from Window->WindowPartition.

See https://github.com/facebookincubator/velox/pull/9266/commits/355e13f535114eb360e9343ae78d1a309e5f1fdd for changes over some of the code in this fix.

aditi-pandit commented on velox/exec/WindowBuild.h:
```diff
@@ -76,6 +76,10 @@ class WindowBuild {
     return data_->estimateRowSize();
   }
 
+  column_index_t inputChannelOf(column_index_t originalIndex) const {
```
I don't think this function is needed after the changes in https://github.com/facebookincubator/velox/pull/9266

aditi-pandit commented on velox/functions/prestosql/aggregates/tests/SumTest.cpp:
```diff
@@ -462,5 +462,44 @@ TEST_F(SumTest, distinct) {
       .assertResults("SELECT c0, sum(distinct c1) FROM tmp GROUP BY 1");
 }
 
+TEST_F(SumTest, aaa) {
```
Change the test name.

kagamiori commented:
Hi @aditi-pandit, thanks a lot for the suggestions in #9266! I've updated the PR and addressed your comments. I've also added addition comments about the expected order of columns in WindowBuild::data_ and WindowBuild::inputColumns_ for clarity. I've combined the computation of inversedInputChannels_ with the computation of inputChannels_ and inputType_ together. Please let me know if you would like a different refactoring or have further suggestions. Thanks!

aditi-pandit commented on velox/exec/WindowBuild.h:
```diff
@@ -91,22 +91,28 @@ class WindowBuild {
-  // RowColumns for window build used to construct WindowPartition.
+  // RowColumns for window build used to construct WindowPartition. Columns are
+  // in the same order as the original input before column reordering.
   std::vector<exec::RowColumn> inputColumns_;
```
Is this variable needed anymore ? We can remove it if so.

aditi-pandit reviewed (commented):
Thanks @kagamiori. This is looking very good. One minor comment.

aditi-pandit commented on velox/functions/prestosql/aggregates/tests/SumTest.cpp:
```diff
@@ -462,5 +462,66 @@ TEST_F(SumTest, distinct) {
       .assertResults("SELECT c0, sum(distinct c1) FROM tmp GROUP BY 1");
 }
 
+TEST_F(SumTest, kRangeAndKRowsWithUnorderedKeyColumns) {
```
Can this test be moved to AggregateWindowTest.cpp ?

kagamiori commented:
Hi @aditi-pandit, I've updated the PR to address your comments. Please take a look again. Thanks!

aditi-pandit reviewed (approved):
Thanks @kagamiori
